File: jas/tools/yaml_tool_pen_test.py
"""Pen-tool behavioral tests — Python port of the Rust pen seam tests in
jas_dioxus/src/tools/yaml_tool.rs (the ``pen_parity_*`` family) and the Swift
port in JasSwift/Tests/Tools/YamlToolPenTests.swift.

These cover the externally-observable outcomes of the YAML-driven pen tool
loaded from the compiled workspace bundle: click-click-click creates a
polyline, click-drag sets the out-handle, click-near-first closes the path,
double-click commits open, and Escape either commits (>= 2 anchors) or
discards (< 2 anchors).

Seam mapping from Rust to this app (Python / PySide6):
    on_press        -> on_press(ctx, x, y, shift, alt)
    on_move(drag)   -> on_move(ctx, x, y, shift, alt, dragging)
    on_release      -> on_release(ctx, x, y, shift, alt)
    on_double_click -> on_double_click(ctx, x, y)   (dispatches on_dblclick)
    on_key_event    -> on_key_event(ctx, key, mods) (dispatches on_keydown)

ESCAPE ENTRY POINT (the Rust on_key vs on_key_event regression guard).
Rust's app shell calls ``tool.on_key`` for Escape/Enter, so a YamlTool that
only overrode ``on_key_event`` would miss Escape (the dx-serve bug that
surfaced the guard). In THIS app the split is different but the conclusion is
the same as Swift's: ``CanvasTool.on_key`` is the int-keyCode handler and
``YamlTool`` does NOT override it (it returns False — see
tools/yaml_tool.py on_key). The string ``on_key_event`` is the handler
YamlTool actually implements (it dispatches the spec's ``on_keydown``). The
canvas's real Escape path for a NON-CAPTURING tool like the pen is, in
canvas/canvas.py keyPressEvent: ``on_key`` is tried first (returns False for
the pen), then the Escape branch calls
``self._active_tool.on_key_event(self._tool_ctx, "Escape", mods)``
(canvas.py ~line 2238). So in Python the Rust on_key regression guard and the
on_key_event path collapse onto the SAME method, ``on_key_event``. Test #4
drives that exact entry point — the one the canvas dispatches — which is the
equivalent regression guard.

The committed element is a Path whose ``d`` is a tuple of path commands
(MoveTo / LineTo / CurveTo / ClosePath dataclasses from geometry.element).
Each test asserts on the committed Path at ``layers[0].children[0]``.
"""

from __future__ import annotations

import json
import os
import sys

_REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
if _REPO_ROOT not in sys.path:
    sys.path.insert(0, _REPO_ROOT)
_JAS_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
if _JAS_DIR not in sys.path:
    sys.path.insert(0, _JAS_DIR)

import pytest

from document.controller import Controller
from document.document import Document
from document.model import Model
from geometry.element import (
    ClosePath, CurveTo, Layer, MoveTo, Path,
)
from tools.tool import KeyMods, ToolContext
from tools.yaml_tool import YamlTool


# ── Loader + harness helpers ───────────────────────────────


def _pen_tool() -> YamlTool | None:
    """Load the "pen" spec from the compiled workspace bundle and build a
    YamlTool, mirroring the selection-tool loader in yaml_tool_test.py."""
    ws_path = os.path.abspath(os.path.join(
        _REPO_ROOT, "workspace", "workspace.json",
    ))
    if not os.path.exists(ws_path):
        return None
    with open(ws_path, "r") as f:
        data = json.load(f)
    tools = data.get("tools")
    if not isinstance(tools, dict):
        return None
    spec = tools.get("pen")
    return YamlTool.from_workspace_tool(spec) if spec else None


def _empty_layer_model() -> Model:
    """A document with a single empty top-level layer — the empty-layer model
    the Rust/Swift ports build (committed paths land at layers[0].children[0])."""
    layer = Layer(name="L", children=())
    return Model(document=Document(layers=(layer,)))


def _ctx(model: Model) -> ToolContext:
    """A faithful headless ToolContext over a live document, with inert
    hit-test callbacks (the pen tool does no hit-testing). Mirrors the
    cross_language_test.py gesture harness ToolContext."""
    return ToolContext(
        model=model,
        controller=Controller(model=model),
        hit_test_selection=lambda x, y: False,
        hit_test_handle=lambda x, y: None,
        hit_test_text=lambda x, y: None,
        hit_test_path_curve=lambda x, y: None,
        request_update=lambda: None,
    )


def _click(tool: YamlTool, ctx: ToolContext, x: float, y: float) -> None:
    """A single pen "click": press + release at the same point, no modifiers."""
    tool.on_press(ctx, x, y, False, False)
    tool.on_release(ctx, x, y, False, False)


def _committed_path(model: Model) -> Path:
    children = model.document.layers[0].children
    assert len(children) == 1, f"expected exactly one child, got {len(children)}"
    el = children[0]
    assert isinstance(el, Path), f"expected Path, got {type(el).__name__}"
    return el


class TestPenTool:
    # ── Loader sanity ──────────────────────────────────────

    def test_pen_tool_loads_from_workspace(self):
        tool = _pen_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        assert tool.spec.id == "pen"

    # ── 1. Three clicks + double-click -> open polyline ────

    def test_three_clicks_then_double_click_creates_polyline(self):
        tool = _pen_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        # Click, click, click — each mouseup lands mode=placing and leaves the
        # anchor in the buffer. Handles stay at anchor position (corner anchors).
        _click(tool, ctx, 10, 10)
        _click(tool, ctx, 50, 10)
        _click(tool, ctx, 50, 50)
        # Double-click (the second press pushed a fourth anchor; the dblclick
        # handler pops it, leaving 3 anchors), then double-clicks at (50,50).
        _click(tool, ctx, 50, 50)
        tool.on_double_click(ctx, 50, 50)

        pe = _committed_path(model)
        # MoveTo + 2 CurveTos (3 anchors -> 2 segments). No ClosePath because
        # dblclick commits open.
        assert len(pe.d) == 3
        assert isinstance(pe.d[0], MoveTo)
        assert pe.d[0].x == 10.0 and pe.d[0].y == 10.0
        assert isinstance(pe.d[1], CurveTo)
        assert isinstance(pe.d[2], CurveTo)
        assert not isinstance(pe.d[-1], ClosePath), (
            "dblclick should commit OPEN, but last command is ClosePath")

    # ── 2. Click-drag sets out-handle ──────────────────────

    def test_click_drag_sets_out_handle(self):
        tool = _pen_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        # First anchor: click + drag out to (60, 10). on_move sets the handle;
        # the first anchor's out = (60, 10), in mirrors to (-40, 10).
        tool.on_press(ctx, 10, 10, False, False)
        tool.on_move(ctx, 60, 10, False, False, True)
        tool.on_release(ctx, 60, 10, False, False)
        # Second anchor: plain click at (50, 50).
        _click(tool, ctx, 50, 50)
        # Escape commits open — exercising the on_keydown path too.
        tool.on_key_event(ctx, "Escape", KeyMods())

        pe = _committed_path(model)
        # d[0] = MoveTo(10,10); d[1] = CurveTo(prev_out=(60,10), curr_in=(50,50),
        # curr=(50,50)) because the second anchor is a corner.
        assert len(pe.d) == 2
        c = pe.d[1]
        assert isinstance(c, CurveTo)
        assert c.x1 == 60.0, "prev anchor out-handle x"
        assert c.y1 == 10.0, "prev anchor out-handle y"
        assert c.x == 50.0
        assert c.y == 50.0

    # ── 3. Click near first anchor closes ──────────────────

    def test_click_near_first_closes(self):
        tool = _pen_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        # Three corner anchors.
        _click(tool, ctx, 10, 10)
        _click(tool, ctx, 50, 10)
        _click(tool, ctx, 50, 50)
        # Fourth click within 8 px of the first anchor (10, 10).
        _click(tool, ctx, 11, 11)

        pe = _committed_path(model)
        # Should end with ClosePath.
        assert isinstance(pe.d[-1], ClosePath), (
            f"expected last command ClosePath, got {type(pe.d[-1]).__name__}")

    # ── 4. Escape via the canvas's actual key path ─────────

    def test_escape_via_shell_key_path_commits(self):
        # Regression guard for the canvas keyboard path. The Rust shell calls
        # tool.on_key() (NOT on_key_event); a YamlTool overriding only
        # on_key_event would miss Escape (the dx-serve bug that surfaced this).
        # This app's canvas dispatches Escape for a NON-CAPTURING tool via
        # canvas.py keyPressEvent -> on_key (returns False for the pen) -> the
        # Escape branch -> on_key_event(ctx, "Escape", mods) (canvas.py ~2238).
        # So on_key_event is exactly the entry point the canvas drives here.
        tool = _pen_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        _click(tool, ctx, 10, 10)
        _click(tool, ctx, 50, 50)
        # The canvas's actual Escape dispatch for a non-capturing tool.
        tool.on_key_event(ctx, "Escape", KeyMods())

        pe = _committed_path(model)
        assert isinstance(pe.d[0], MoveTo)
        assert not isinstance(pe.d[-1], ClosePath), (
            "Escape should commit OPEN, but last command is ClosePath")

    # ── 5. Escape with a single anchor discards ────────────

    def test_escape_without_enough_anchors_discards(self):
        tool = _pen_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        # One anchor — not enough to make a path (commit gates on
        # anchor_buffer_length('pen') >= 2).
        _click(tool, ctx, 10, 10)
        tool.on_key_event(ctx, "Escape", KeyMods())
        assert len(model.document.layers[0].children) == 0, (
            "single anchor should not produce a path")

File: jas_ocaml/test/tools/yaml_tool_pen_test.ml
(** Pen-tool behavioral tests — OCaml port of the Rust pen seam tests
    in jas_dioxus/src/tools/yaml_tool.rs (the pen_parity_* family) and
    the Swift port in JasSwift/Tests/Tools/YamlToolPenTests.swift.

    These cover the externally-observable outcomes of the YAML-driven
    pen tool loaded from the workspace bundle: click-click-click creates
    a polyline, click-drag sets the out-handle, click-near-first closes,
    double-click commits open, and Escape either commits (anchors >= 2)
    or discards (anchors < 2).

    Seam mapping from Rust to OCaml:
      on_press        -> on_press ctx x y ~shift ~alt
      on_move(drag)   -> on_move ctx x y ~shift ~alt ~dragging
      on_release      -> on_release ctx x y ~shift ~alt
      on_double_click -> on_double_click ctx x y
      on_key_event    -> on_key_event ctx key mods

    Escape entry point. Rust distinguishes tool.on_key (the app shell
    Escape/Enter call path) from tool.on_key_event, and keeps a regression
    guard exercising on_key. In OCaml the canvas_tool method [on_key]
    takes an [int] keycode and the YamlTool does not handle Escape there;
    the string-based handler is [on_key_event]. The GTK canvas shell, in
    Canvas_subwindow.forward_key_event, builds a string key name and for a
    non-capturing tool routes Escape and Enter through
    [active_tool#on_key_event ctx k mods] (see the branch guarded by
    [captures_keyboard () || k = "Escape" || k = "Enter"]). So the OCaml
    shell Escape path for a non-capturing tool like the pen collapses onto
    [on_key_event], exactly as in Swift. pen_parity_escape_via_shell_key_path
    therefore drives the exact entry the canvas dispatches, which is the
    equivalent regression guard. *)

open Jas

let () = ignore (GMain.init ())

let pen_tool () : Yaml_tool.yaml_tool option =
  match Workspace_loader.load () with
  | None -> None
  | Some ws ->
    match Workspace_loader.json_member "tools" ws.data with
    | Some (`Assoc tools) ->
      (match List.assoc_opt "pen" tools with
       | Some spec -> Yaml_tool.from_workspace_tool spec
       | None -> None)
    | _ -> None

(** Document with a single empty layer (no children). *)
let empty_layer_model () : Model.model =
  let layer = Element.Layer {
    name = Some "L";
    id = None;
    children = [||];
    transform = None; locked = false; opacity = 1.0;
    visibility = Preview; blend_mode = Normal; mask = None;
    isolated_blending = false; knockout_group = false;
  } in
  let m = Model.create () in
  m#set_document_unbracketed (Document.make_document [| layer |]);
  m

let make_ctx model =
  let ctrl = new Controller.controller ~model () in
  let ctx : Canvas_tool.tool_context = {
    model;
    controller = ctrl;
    hit_test_selection = (fun _ _ -> false);
    hit_test_handle = (fun _ _ -> None);
    hit_test_text = (fun _ _ -> None);
    hit_test_path_curve = (fun _ _ -> None);
    request_update = (fun () -> ());
    draw_element_overlay = (fun _cr _elem ~is_partial:_ _cps -> ());
  } in
  (ctx, ctrl)

let no_mods : Canvas_tool.key_mods =
  { shift = false; ctrl = false; alt = false; meta = false }

(** A single pen click: press + release at the same point, no modifiers. *)
let click (tool : Yaml_tool.yaml_tool) ctx x y =
  tool#on_press ctx x y ~shift:false ~alt:false;
  tool#on_release ctx x y ~shift:false ~alt:false

(** Children of the first layer. *)
let layer0_children (m : Model.model) : Element.element array =
  match m#document.layers.(0) with
  | Element.Layer { children; _ } -> children
  | _ -> [||]

(** The d command list of the committed path at layers[0].children[0]. *)
let committed_d (m : Model.model) : Element.path_command list =
  match (layer0_children m).(0) with
  | Element.Path { d; _ } -> d
  | _ -> assert false

let last lst = List.nth lst (List.length lst - 1)

(* ── Loader sanity ─────────────────────────────────── *)

let load_tests = [
  Alcotest.test_case "pen_tool_loads_from_workspace" `Quick (fun () ->
    match pen_tool () with
    | Some tool ->
      assert (tool#spec.id = "pen")
    | None -> Alcotest.skip ());
]

(* ── Three clicks + double-click -> open polyline ───── *)

let polyline_tests = [
  Alcotest.test_case
    "three_clicks_then_double_click_creates_polyline" `Quick (fun () ->
    match pen_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      (* Click, click, click — each mouseup lands mode=placing and
         leaves the anchor in the buffer. Handles stay at anchor
         position (corner anchors). *)
      click tool ctx 10.0 10.0;
      click tool ctx 50.0 10.0;
      click tool ctx 50.0 50.0;
      (* Double-click (the second press pushed a fourth anchor; the
         dblclick handler pops it, leaving 3 anchors). *)
      click tool ctx 50.0 50.0;
      tool#on_double_click ctx 50.0 50.0;

      let children = layer0_children m in
      assert (Array.length children = 1);
      let d = committed_d m in
      (* MoveTo + 2 CurveTos (3 anchors -> 2 segments). No ClosePath
         because dblclick commits open. *)
      assert (List.length d = 3);
      (match List.nth d 0 with
       | Element.MoveTo (x, y) -> assert (x = 10.0 && y = 10.0)
       | _ -> assert false);
      (match List.nth d 1 with
       | Element.CurveTo _ -> ()
       | _ -> assert false);
      (match List.nth d 2 with
       | Element.CurveTo _ -> ()
       | _ -> assert false);
      (match last d with
       | Element.ClosePath -> assert false
       | _ -> ()));
]

(* ── Click-drag sets out-handle ─────────────────────── *)

let drag_handle_tests = [
  Alcotest.test_case "click_drag_sets_out_handle" `Quick (fun () ->
    match pen_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      (* First anchor: click + drag out to (60, 10). on_move sets the
         handle; the first anchor out = (60, 10), in mirrors to
         (-40, 10). *)
      tool#on_press ctx 10.0 10.0 ~shift:false ~alt:false;
      tool#on_move ctx 60.0 10.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_release ctx 60.0 10.0 ~shift:false ~alt:false;
      (* Second anchor: plain click at (50, 50). *)
      click tool ctx 50.0 50.0;
      (* Escape commits open — exercising the on_keydown path too. *)
      let _ = tool#on_key_event ctx "Escape" no_mods in

      let children = layer0_children m in
      assert (Array.length children = 1);
      let d = committed_d m in
      (* d[0] = MoveTo(10,10); d[1] = CurveTo(prev_out=(60,10),
         curr_in=(50,50), curr=(50,50)) because the second anchor is a
         corner. *)
      assert (List.length d = 2);
      (match List.nth d 1 with
       | Element.CurveTo (x1, y1, _, _, x, y) ->
         assert (x1 = 60.0);
         assert (y1 = 10.0);
         assert (x = 50.0);
         assert (y = 50.0)
       | _ -> assert false));
]

(* ── Click near first anchor closes ─────────────────── *)

let close_tests = [
  Alcotest.test_case "click_near_first_closes" `Quick (fun () ->
    match pen_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      (* Three corner anchors. *)
      click tool ctx 10.0 10.0;
      click tool ctx 50.0 10.0;
      click tool ctx 50.0 50.0;
      (* Fourth click within 8 px of the first anchor (10, 10). *)
      click tool ctx 11.0 11.0;

      let children = layer0_children m in
      assert (Array.length children = 1);
      let d = committed_d m in
      (* Should end with ClosePath. *)
      (match last d with
       | Element.ClosePath -> ()
       | _ -> assert false));
]

(* ── Escape via the app shell key path ──────────────── *)

let escape_commit_tests = [
  Alcotest.test_case "escape_via_shell_key_path_commits" `Quick (fun () ->
    (* Regression guard for the canvas keyboard path. The Rust shell
       calls tool.on_key (NOT on_key_event); a YamlTool overriding only
       on_key_event would miss Escape (the dx-serve bug that surfaced
       this). The OCaml shell equivalent for a non-capturing tool is
       Canvas_subwindow.forward_key_event ->
       active_tool#on_key_event ctx "Escape" mods, which is exactly what
       we drive here. *)
    match pen_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      click tool ctx 10.0 10.0;
      click tool ctx 50.0 50.0;
      (* The shell actual Escape dispatch for a non-capturing tool. *)
      let _ = tool#on_key_event ctx "Escape" no_mods in

      let children = layer0_children m in
      assert (Array.length children = 1);
      let d = committed_d m in
      (match List.nth d 0 with
       | Element.MoveTo _ -> ()
       | _ -> assert false);
      (match last d with
       | Element.ClosePath -> assert false
       | _ -> ()));
]

(* ── Escape with a single anchor discards ───────────── *)

let escape_discard_tests = [
  Alcotest.test_case
    "escape_without_enough_anchors_discards" `Quick (fun () ->
    match pen_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      (* One anchor — not enough to make a path. *)
      click tool ctx 10.0 10.0;
      let _ = tool#on_key_event ctx "Escape" no_mods in
      assert (Array.length (layer0_children m) = 0));
]

let () =
  Alcotest.run "Yaml pen tool" [
    "Tool load", load_tests;
    "Polyline", polyline_tests;
    "Click-drag handle", drag_handle_tests;
    "Close", close_tests;
    "Escape commit", escape_commit_tests;
    "Escape discard", escape_discard_tests;
  ]
